app/scripts/migrations/092.3.ts

```diff
@@ -0,0 +1,92 @@
+import { isNullOrUndefined, isObject } from '@metamask/utils';
+import { cloneDeep } from 'lodash';
+import log from 'loglevel';
+
+type VersionedData = {
+  meta: { version: number };
+  data: Record<string, unknown>;
+};
+
+export const version = 92.3;
+
+/**
+ * This migration does the following:
+ *
+ * - Deletes currently stored advancedGasFee in preferences controller,
+ * replacing the default with an empty object
+ * - Sets hadAdvancedGasFeesSetPriorToMigration92_3 flag on AppStateController
+ * to indicate if the user had previously had advancedGasFee set in their
+ * preferences. This will be used to display a whats new entry to inform users
+ * that we wiped these settings and made them apply per network.
+ *
+ * @param originalVersionedData - Versioned MetaMask extension state, exactly what we persist to dist.
+ * @param originalVersionedData.meta - State metadata.
+ * @param originalVersionedData.meta.version - The current state version.
+ * @param originalVersionedData.data - The persisted MetaMask state, keyed by controller.
+ * @returns Updated versioned MetaMask extension state.
+ */
+export async function migrate(
+  originalVersionedData: VersionedData,
+): Promise<VersionedData> {
+  const versionedData = cloneDeep(originalVersionedData);
+  versionedData.meta.version = version;
+  migrateData(versionedData.data);
+  return versionedData;
+}
+
+function migrateData(state: Record<string, unknown>): void {
+  changeShapeAndRemoveOldAdvancedGasFeePreference(state);
+}
+
+function changeShapeAndRemoveOldAdvancedGasFeePreference(
+  state: Record<string, any>,
+) {
+  if (isNullOrUndefined(state.PreferencesController)) {
+    log.warn(
+      `Migration #${version}: preferences controller null or undefined, skipping migration`,
+    );
+    return;
+  }
+  const possibleOriginalValue = state.PreferencesController?.advancedGasFee;
+
+  // Will be false if the keys set on the object are anything other than the
+  // maxBaseFee or priorityFee. Essentially if the object is already keyed
+  // by chainId it won't show as hadFeesSet.
+  const hadFeesSet =
+    isObject(possibleOriginalValue) &&
+    hasFeePreferenceKeys(possibleOriginalValue);
+
+  if (isObject(state.AppStateController) === false) {
+    global.sentry?.captureException?.(
+      new Error(
+        `typeof state.AppStateController is ${typeof state.AppStateController}`,
+      ),
+    );
+  } else {
+    state.AppStateController.hadAdvancedGasFeesSetPriorToMigration92_3 =
+      hadFeesSet;
+  }
+
+  if (isObject(state.PreferencesController) === false) {
```

Comment:
Typically these "corrupted state" checks are done prior to any changes to state. Maybe that would be better here to? Seems easier to reason about the corrupted state scenario in that case, no need to worry about a partial migration. This check helps with type narrowing as well

Reply:
Yeah, as I was doing the unknown change above I realized I kind of had to do this or else have fairly verbose code anyways. It doesn't make sense to have a partial migration... the key in either would be in a weird out of sync state if either of them completed successfully.